src/main/java/dev/jbang/cli/BaseScriptCommand.java

```diff
@@ -155,49 +155,50 @@ public static Script prepareScript(String scriptResource, List<String> arguments
 			s.setOriginal(scriptResource);
 			s.setAdditionalDependencies(dependencies);
 			s.setAdditionalClasspaths(classpaths);
-
-			// Logic to get all //SOURCES recursively
-			List<Path> resolvedSourcePaths = new ArrayList<>();
-			// It's important to know which sources we already visited,
-			// because many files can declare the same source.
-			Set<String> visited = new HashSet<>();
-			// List of array: [0] original source [1] source
-			List<String[]> sources = new ArrayList<>();
-			// Collect sources from the entry point (main file)
-			String originalSource = s.getScriptResource().getOriginalResource();
-			List<FileRef> fileRefs = s.collectSources();
-			for (FileRef fileRef : fileRefs)
-				sources.add(new String[] { originalSource, fileRef.getDestination() });
-			while (!sources.isEmpty()) {
-				String[] tmp = sources.remove(0);
-				originalSource = tmp[0];
-				String source = tmp[1];
-				if (!visited.add(source))
-					continue;
-				source = Util.swizzleURL(source); // base path for new sources
-				Path path = s.getScriptResource().fetchIfNeeded(source, originalSource);
-				resolvedSourcePaths.add(path);
-				// TODO would we not be better of with Script ref here rather than a raw string
-				// ?
-				String sourceContent = new String(Files.readAllBytes(path), Charset.defaultCharset());
-				List<String> newSources = Util.collectSources(sourceContent);
-				for (String newSource : newSources) {
-					// If source is a URL then it must be the new base path
-					if (Util.isURL(source)) {
-						sources.add(new String[] { source, newSource });
-					} else {
-						sources.add(new String[] { originalSource, newSource });
-					}
-				}
-			}
-
-			s.setResolvedSources(resolvedSourcePaths);
+			s.setResolvedSources(resolveSOURCESRecursively(s));
 		} catch (FileNotFoundException e) {
 			throw new ExitException(1, e);
 		}
 		return s;
 	}
 
+	public static List<Path> resolveSOURCESRecursively(Script script) throws IOException {
```

Comment:
this does not seem related to this fix ?

Reply:
This one you didn't merge from previous PR. You also didn't merge the test.
I extracted to a function to make it easier to test. But it end up making the code more readable in my opinion.